### PR TITLE
perf: count(*) for append-only tables 

### LINKED ISSUE #4516: Optimize count(*) for append-only tables
### What type of enhancement is this?

Performance

### What does the enhancement do?

Since log tables are always "append-only", we can summarize total rows by simply adding all rows in memtables/sst files instead of scanning all rows.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- closes #4516

## What's changed and what's your intention?

Optimize `count(*)` qeuries for tables with append mode enabled:

Test result on TSBS dataset (103680000 rows, append_mode=true): ~600x faster

```
# Before (approx. 4160ms)
mysql> select count(*) from benchmark.cpu;
+-----------+
| COUNT(*)  |
+-----------+
| 103680000 |
+-----------+
1 row in set (4.21 sec)

# After (approx. 7ms)
mysql> select count(*) from benchmark.cpu;
+-----------+
| COUNT(*)  |
+-----------+
| 103680000 |
+-----------+
1 row in set (0.01 sec)
```

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable.rs [hunk 1]
```diff
@@ -69,6 +69,8 @@ pub struct MemtableStats {
     /// The time range that this memtable contains. It is None if
     /// and only if the memtable is empty.
     time_range: Option<(Timestamp, Timestamp)>,
+    /// Total rows in memtable
+    num_rows: usize,
```
Comment (evenyag) [inline src/mito2/src/memtable.rs:73]: We should document that it also contains the deleted rows.

File: src/mito2/src/read/scan_region.rs [hunk 2]
```diff
@@ -697,6 +697,12 @@ impl ScanInput {
             }
         });
     }
+
+    pub(crate) fn total_rows(&self) -> usize {
+        let rows_in_files: usize = self.files.iter().map(|f| f.num_rows()).sum();
+        let rows_in_memtables: usize = self.memtables.iter().map(|m| m.stats().num_rows()).sum();
```
Comment (evenyag) [inline src/mito2/src/read/scan_region.rs:703]: We might filter memtables by time range in `ScanRegion::scan_input()`. Will it affect the correctness?

File: src/mito2/src/read/seq_scan.rs [hunk 3]
```diff
@@ -65,7 +65,11 @@ impl SeqScan {
     /// Creates a new [SeqScan].
     pub(crate) fn new(input: ScanInput) -> Self {
         let parallelism = input.parallelism.parallelism.max(1);
-        let properties = ScannerProperties::new_with_partitions(parallelism);
+        let properties = ScannerProperties {
+            partitions: vec![vec![]; parallelism],
+            append_mode: input.append_mode,
+            total_rows: input.total_rows(),
+        };
```
Comment (evenyag) [inline src/mito2/src/read/seq_scan.rs:72]: non-blocking: Maybe we can add two functions `with_append_mode()` and `with_total_rows()` so we can avoid repeating `vec![vec![]; parallelism]`.

File: src/store-api/src/region_engine.rs [hunk 4]
```diff
@@ -164,18 +164,19 @@ pub struct ScannerProperties {
     /// The first dim vector's length represents the output partition number. The second
     /// dim is ranges within one partition.
     pub partitions: Vec<Vec<PartitionRange>>,
+
+    pub append_mode: bool,
+
+    pub total_rows: usize,
```
Comment (evenyag) [inline src/store-api/src/region_engine.rs:170]: nitpick: Document the relationship between these two fields.

File: src/table/src/table/scan.rs [hunk 5]
```diff
@@ -152,6 +159,28 @@ impl ExecutionPlan for RegionScanExec {
...
     }
+
+    fn statistics(&self) -> DfResult<Statistics> {
+        let statistics = if self.append_mode {
+            let column_statistics = self
+                .arrow_schema
+                .fields
+                .iter()
+                .map(|_| ColumnStatistics {
+                    distinct_count: Precision::Exact(self.total_rows),
```
Comment (evenyag) [inline src/table/src/table/scan.rs:170]: The distinct count may not be `total_rows` as each column may have duplicated values.

Comment (v0y4g3r) [reply src/table/src/table/scan.rs:170]: > The distinct count may not be `total_rows` as each column may have duplicated values.

`ColumnStatistics` is only read iff statement is `select count(*) from table` (without predicates) and `append_mode` is true. Maybe I should add some doc here.

File: src/mito2/src/read/scan_region.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/read/scan_region.rs:703]: > We might filter memtables by time range in `ScanRegion::scan_input()`. Will it affect the correctness?

No, as I commented [here](https://github.com/GreptimeTeam/greptimedb/pull/4545#discussion_r1713643681)

File: src/common/query/Cargo.toml [hunk 6]
```diff
@@ -21,6 +22,7 @@ common-time.workspace = true
 datafusion.workspace = true
 datafusion-common.workspace = true
 datafusion-expr.workspace = true
+datafusion-physical-expr.workspace = true
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/common/query/Cargo.toml]: fixed in 92e5295

File: src/mito2/src/memtable.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable.rs:73]: fixed in 92e5295

File: src/mito2/src/read/seq_scan.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/mito2/src/read/seq_scan.rs]: fixed in 92e5295

File: src/store-api/src/region_engine.rs (hunk 4 above)
Comment (v0y4g3r) [reply src/store-api/src/region_engine.rs]: fixed in 92e5295

File: src/table/src/table/scan.rs (hunk 5 above)
Comment (evenyag) [reply src/table/src/table/scan.rs:170]: Maybe we can document it in this method.

**Review by evenyag — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**